creusot/tests/should_succeed/mapping_test.rs

```diff
@@ -20,16 +18,16 @@ impl Model for T {
     fn model(self) -> Self::ModelTy {
         std::process::abort()
     }
-
 }
 
 #[requires( 0 <= @((*t).a) )] // otherwise its wrong !
 #[requires( @((*t).a) < 1000 )] // to prevent overflow
-#[ensures( (@^t).eq((@*t).set(@((*t).a),1)) )]
-fn incr(t:&mut T) {
+#[ensures( @^t === (@*t).set(@((*t).a),1) )]
```

Comment:
adding `LogicalEq` means you can use `==`. this is key since your equality needs to be extensional unlike `===`.

Reply:
Ah, interesting, I would have figured it would be better to consistently use `==`

Reply:
Doesn't the definition of `==` have something like `a == b -> a = b`?

Reply:
Note that I would not expect `===` and `==` to be equivalent for mappings.

`===` should be logical, Leibniz equality, while `==` should be pointwise `==` equality. If `===` and `==` are not equivalent on the underlying type, then `===` and `==` should not be equivalent (even if mappings are extensional).

This is what I would expect from the notations. Question is: is this really what we want? In particular, this means that we do a different distinction between `===` and `==` than Why3 is doing...

Reply:
We do have extensionality for `==` on mappings:

```
lemma extensionality:
    forall m1 m2: 'a -> 'b. m1 == m2 -> m1 = m2
```

Where `==` is defined pointwise using `=`. 

> This is what I would expect from the notations. Question is: is this really what we want? In particular, this means that we do a different distinction between === and == than Why3 is doing...

That is the real question, and we actually _can_ do something different (perhaps even *better*) as we can rely on traits. Why3 can't / doesn't parameterize definitions of `==` by the `==` on the underlying type, but we can.

Reply:
But I had the idea of defining `===` as a default for `LogicalEq` and using specialization to allow overriding it with more specific notions of equality but is that appropriate?

Reply:
```
impl<A, B: LogicalEq> LogicalEq for Map<A, B> { 
  #[predicate]
  fn eq_log(self, o: Self) -> bool {
    forall<a : A> self.get(a) == o.get(a)
  }
}
```

Reply:
Yeah, that's what I would expect. But then this does not meet the notion in Why3, and perhaps we will not leverage the lemmas that are proved in Why3's standard library about its `==`...

Reply:
> leverage the lemmas that are proved in Why3's standard library about its ==...

There's only one and given the comment in the library, i'm not even sure it is proven in the first place.

> In short, === (that is = in Why3) is built-in in provers, whereas == is not, that's why I'd like to use === whenever possible.
> And there is the fact that extensionality lemmas like to one for mappings above is not applied easily by provers (where as the converse is a triviality)

This makes sense, perhaps the 'idiom' would be to use `==` inside a function and `===` in its external spec? Or could there be ways to better hint to provers to more systematically attempt extensionality?

Reply:
> In the Why3 logic, and its theory of mappings, === and == are proved equivalent.

That is not true if the equality on which `==` is defined is not the logical equality over elements.

Reply:
> That is not true if the equality on which == is defined is not the logical equality over elements.

Should we make extensionality a required law of `LogicalEq` ?

Reply:
That would mean making `===` and `==` equivalent for every type in Creusot.

I am not so sure this is what we want. This would really defeat the point of `LogicalEq`: this trait is here to define an alternative notion of equality if needed.

Reply:
> This would really defeat the point of LogicalEq: this trait is here to define an alternative notion of equality if needed.

Not necessarily defeat, as it allows us to sidestep the intensionality of `=`, for example: on `seq`, you often don't want `=` as you find yourself incapable to prove simple properties when you build / modify the sequence. That's why why3 comes with a definition of `==` which is extensional.